Add Regex Filter for messages

celestix opened the pull request:
<!--
Hey, thank you for opening a PR!

Please fill out the details below; they're there to help both you and the maintainers!
-->

# What

Added regex filters to allow message handlers to match regex patterns.

Commit:
Added Regex Filter

PaulSonOfLars commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,6 +47,16 @@ func ForwardFromChatID(id int64) filters.Message {
 
+func Regex(pattern string) filters.Message {
+	return func(m *gotgbot.Message) bool {
+		matched, err := regexp.MatchString(pattern, m.Text)
```
Regexes that are reused should be compiled for speed, see `regex.Compile`

ALiwoto replied:
Agree

celestix replied:
Okay, Understood.

PaulSonOfLars commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,6 +47,16 @@ func ForwardFromChatID(id int64) filters.Message {
+	return func(m *gotgbot.Message) bool {
+		matched, err := regexp.MatchString(pattern, m.Text)
+		if err != nil {
+			return false
```
This swallows all errors, making it very hard to debug. How do you know your match failed?

One option would be to change the definition to `Regex(p string) (filters.Message, error)`. This stops users from being able to instantiate the filter inline, but still allows them to handle the error.

celestix replied:
I see, going to try it out today. :) Thanks for the brief explanation.

ALiwoto replied:
Why not changing the type `filters.Message` from `func(*gotgbot.Message) bool` to `func(*gotgbot.Message) (bool, error)`?
Isn't it a good idea?

ALiwoto replied:
~@anonyindian  please check my pull request [here](https://github.com/anonyindian/gotgbot/pull/1).~

PaulSonOfLars replied:
The compile error only ever happens once (at compile time). The error can also only ever be handled at compile time, so there's no way to action any errors that would get returned with (bool, error) - so, what do you gain by returning the error?

Commit:
Added Regex Filters

- Compiling regex pattern before match should increase its speed.
- log.fatal should accomplish error handling issue.

celestix commented:
Just committed both the changes. once have a look @PaulSonOfLars

ALiwoto commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,12 +49,14 @@ func ForwardFromChatID(id int64) filters.Message {
-		matched, err := regexp.MatchString(pattern, m.Text)
+		r, err := regexp.Compile(p)
 		if err != nil {
+			log.Fatal(err)
```
Oh god
no please don't do that

ALiwoto replied:
`log.Fatal(err)` 
this work is completely wrong
if we get an error, the application will completely terminated
it's not called error handling at all

PaulSonOfLars commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,6 +49,18 @@ func ForwardFromChatID(id int64) filters.Message {
+	return func(m *gotgbot.Message) bool {
+		r, err := regexp.Compile(p)
+		if err != nil {
+			log.Fatal(err)
```
This will cause the program to panic. That is definitely not the intention here. The programmer needs to handle the error that gets returned - if they don't, that's their choice, but killing the program isn't the answer.

That's why I suggested that the `Regex` return both a message, and an error.

PaulSonOfLars commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,6 +49,18 @@ func ForwardFromChatID(id int64) filters.Message {
 
+func Regex(p string) filters.Message {
+	return func(m *gotgbot.Message) bool {
+		r, err := regexp.Compile(p)
```
The compile should be moved outside of the anonymous `func()`, otherwise you recompile the regex every time and lose all the performance improvements from compiling it in the first place.

ALiwoto replied:
Oh right, this way if we got an error from `regexp.Compile`, we can return it by `Regex` function
so the `Regex` function will become something like this `func Regex(p string) (filters.Message, error)`
this way users can handle it directly

celestix replied:
Right, Let me do so.

PaulSonOfLars commented on ext/handlers/filters/message/message.go:
```diff
@@ -47,6 +49,18 @@ func ForwardFromChatID(id int64) filters.Message {
+			log.Fatal(err)
+			return false
+		}
+		matched := r.MatchString(m.Text)
+		return matched
```
```suggestion
	  return r.MatchString(m.Text)
```

No point having a temporary variable here

ALiwoto reviewed (commented):
> so, what do you gain by returning the error?
@PaulSonOfLars  What should be done then? Instead of returning the error?

Commit:
Carried out changed

celestix commented:
@PaulSonOfLars Please review the last commit!

PaulSonOfLars reviewed (approved):
LGTM % last comment to remove an unnecessary temp variable

Commit:
Removed Temporary Variable

celestix commented:
> LGTM % last comment to remove an unnecessary temp variable

Once check please, It should be fine now.

ALiwoto commented on ext/handlers/filters/message/message.go:
```diff
@@ -54,8 +54,7 @@ func Regex(p string) (filters.Message, error) {
 	return func(m *gotgbot.Message) bool {
-		matched := r.MatchString(m.Text)
-		return matched
+		return r.MatchString(m.Text)
```
<img align="middle" src="https://raw.githubusercontent.com/aliwoto/aliwoto/main/resources/Something_that_looks_like_Diamond.png" width="35px" height="35px" ></img> nice
